Framework/MDAlgorithms/inc/MantidMDAlgorithms/IntegrateQLabEvents.h

```diff
@@ -0,0 +1,293 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2012 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#pragma once
+
+#include "MantidDataObjects/EventWorkspace.h"
+#include "MantidDataObjects/Peak.h"
+#include "MantidDataObjects/PeakShapeEllipsoid.h"
+#include "MantidKernel/Matrix.h"
+#include "MantidKernel/V3D.h"
+
+#include <memory>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+namespace Mantid {
+
+namespace Geometry {
+class PeakShape;
+}
+
+namespace MDAlgorithms {
+using Mantid::DataObjects::PeakShapeEllipsoid_const_sptr;
+using Mantid::Geometry::PeakShape_const_sptr;
+using Mantid::Kernel::V3D;
+
+/// Partition QLab space into a cubic lattice
+struct CellCoords {
+  int64_t a;
+  int64_t b;
+  int64_t c;
+
+  CellCoords(const V3D &q, const double cellSize)
+      : a(static_cast<int64_t>(q[0] / cellSize)),
+        b(static_cast<int64_t>(q[1] / cellSize)),
+        c(static_cast<int64_t>(q[2] / cellSize)) {}
+
+  /// Check if all cell coords are zero
+  bool isOrigin() { return !(a || b || c); }
+
+  /// cast coordinates to scalar, to be used as key in an unordered map
+  int64_t getHash() { return 1000000000000 * a + 100000000 * b + 10000 * c; }
+
+  /// Hashes for the 26 first neighbor coordinates plus the coordinates
+  /// themselves
+  std::vector<int64_t> nearbyCellHashes() {
+    std::vector<int64_t> neighbors;
+    for (int64_t ia = a - 1; ia <= a + 1; ia++)
+      for (int64_t ib = b - 1; ib <= b + 1; ib++)
+        for (int64_t ic = c - 1; ic <= c + 1; ic++) {
+          int64_t key = 1000000000000 * ia + 100000000 * ib + 10000 * ic;
+          neighbors.emplace_back(key);
+        }
+    return neighbors;
+  }
+};
+
+// [(weight, error), Q-vector], trimmed-down info for an event
+using SlimEvent = std::pair<std::pair<double, double>, V3D>;
+using SlimEvents = std::vector<SlimEvent>;
+
+// A cell in partitioned QLab space containing one peak
+struct OccupiedCell {
+  size_t peakIndex;  // index of the peak within this cell
+  V3D peakQ;         // QLab vector of the peak within this cell
+  SlimEvents events; // events potentially closer than m_radius to the peak
+};
+
+/**
+ @class IntegrateQLabEvents
+
+ This is a low-level class to construct a map with lists of events near
+ each peak Q-vector in the lab frame. The Q-vector of each event is shifted
+ by the Q-vector of the associated peak.
+ A method is also provided to find the principal axes of such a list
+ of events and to find the net integrated counts using ellipsoids
+ with axis lengths determined from the standard deviations in the
+ directions of the principal axes.
+
+ @author Dennis Mikkelson
+ @date   2012-12-19
+ */
```

Comment:
Some code is new, some code is a duplication of code `Integrate3DEllipsoids.cpp`. I left `Integrate3DEllipsoids.cpp` intact because it's being used by algorithm `IntegrateEllipsoidsTwoSteps`